core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -700,31 +716,18 @@ protected void writerSegmentStats(SegmentsStats stats) {
         ensureOpen();
         Map<String, Segment> segments = new HashMap<>();
         // first, go over and compute the search ones...
-        try (Searcher searcher = acquireSearcher("segments")){
-            for (LeafReaderContext reader : searcher.reader().leaves()) {
-                final SegmentReader segmentReader = Lucene.segmentReader(reader.reader());
-                SegmentCommitInfo info = segmentReader.getSegmentInfo();
-                assert !segments.containsKey(info.info.name);
-                Segment segment = new Segment(info.info.name);
-                segment.search = true;
-                segment.docCount = reader.reader().numDocs();
-                segment.delDocCount = reader.reader().numDeletedDocs();
-                segment.version = info.info.getVersion();
-                segment.compound = info.info.getUseCompoundFile();
-                try {
-                    segment.sizeInBytes = info.sizeInBytes();
-                } catch (IOException e) {
-                    logger.trace((Supplier<?>) () -> new ParameterizedMessage("failed to get size for [{}]", info.info.name), e);
-                }
-                segment.memoryInBytes = segmentReader.ramBytesUsed();
-                segment.segmentSort = info.info.getIndexSort();
-                if (verbose) {
-                    segment.ramTree = Accountables.namedAccountable("root", segmentReader);
-                }
-                segment.attributes = info.info.getAttributes();
-                // TODO: add more fine grained mem stats values to per segment info here
-                segments.put(info.info.name, segment);
-            }
+        try (Searcher searcher = acquireSearcher("segments", SearcherScope.EXTERNAL)){
+            List<SegmentReader> externalSegmentReaders = searcher.reader().getContext()
+                .leaves().stream().map(l -> Lucene.segmentReader(l.reader())).collect(Collectors.toList());
+            fillSegmentInfo(verbose, true, segments, externalSegmentReaders);
+        }
+
+        try (Searcher searcher = acquireSearcher("segments", SearcherScope.INTERNAL)){
+            List<SegmentReader> internalSegmentReaders = searcher.reader().getContext()
+                .leaves().stream().map(l -> Lucene.segmentReader(l.reader()))
+                .filter(r -> segments.containsKey(r.getSegmentName())  == false)
+                .collect(Collectors.toList());
+            fillSegmentInfo(verbose, false, segments, internalSegmentReaders);
```

Comment:
similarly here I would like it better with a regular for loop and by making fillSegmentInfo take a single segment at a time